api/inc/register_gateway.h

```diff
@@ -17,104 +17,146 @@
-#define UVISOR_REGISTER_GATEWAY_MAGIC 0xF7F3A89E
-#else
-#error "Unsupported instruction set. The ARM Thumb-2 instruction set must be supported."
-#endif /* __thumb__ && __thumb2__ */
-
-/* register gateway operations */
-/* note: do not use special characters as these numbers will be stringified */
-#define UVISOR_OP_READ(op)  (op)
-#define UVISOR_OP_WRITE(op) ((1 << 4) | (op))
-#define UVISOR_OP_NOP       0x0
-#define UVISOR_OP_AND       0x1
-#define UVISOR_OP_OR        0x2
-#define UVISOR_OP_XOR       0x3
-
-/* default mask for whole register operatins */
-#define __UVISOR_OP_DEFAULT_MASK 0x0
-
-/* register gateway metadata */
-#if defined(__CC_ARM)
-
-/* TODO/FIXME */
-
-#elif defined(__GNUC__)
-
-/* 1 argument: simple read, no mask */
-#define __UVISOR_REGISTER_GATEWAY_METADATA1(src_box, addr) \
-    "b.n skip_args%=\n" \
-    ".word " UVISOR_TO_STRING(UVISOR_REGISTER_GATEWAY_MAGIC) "\n" \
-    ".word " UVISOR_TO_STRING(addr) "\n" \
-    ".word " UVISOR_TO_STRING(src_box) "_cfg_ptr\n" \
-    ".word " UVISOR_TO_STRING(UVISOR_OP_READ(UVISOR_OP_NOP)) "\n" \
-    ".word " UVISOR_TO_STRING(__UVISOR_OP_DEFAULT_MASK) "\n" \
-    "skip_args%=:\n"
-
-/* 2 arguments: simple write, no mask */
-#define __UVISOR_REGISTER_GATEWAY_METADATA2(src_box, addr, val) \
-    "b.n skip_args%=\n" \
-    ".word " UVISOR_TO_STRING(UVISOR_REGISTER_GATEWAY_MAGIC) "\n" \
-    ".word " UVISOR_TO_STRING(addr) "\n" \
-    ".word " UVISOR_TO_STRING(src_box) "_cfg_ptr\n" \
-    ".word " UVISOR_TO_STRING(UVISOR_OP_WRITE(UVISOR_OP_NOP)) "\n" \
-    ".word " UVISOR_TO_STRING(__UVISOR_OP_DEFAULT_MASK) "\n" \
-    "skip_args%=:\n"
-
-/* 3 arguments: masked read */
-#define __UVISOR_REGISTER_GATEWAY_METADATA3(src_box, addr, op, mask) \
-    "b.n skip_args%=\n" \
-    ".word " UVISOR_TO_STRING(UVISOR_REGISTER_GATEWAY_MAGIC) "\n" \
-    ".word " UVISOR_TO_STRING(addr) "\n" \
-    ".word " UVISOR_TO_STRING(src_box) "_cfg_ptr\n" \
-    ".word " UVISOR_TO_STRING(UVISOR_OP_READ(op)) "\n" \
-    ".word " UVISOR_TO_STRING(mask) "\n" \
-    "skip_args%=:\n"
-
-/* 4 arguments: masked write */
-#define __UVISOR_REGISTER_GATEWAY_METADATA4(src_box, addr, val, op, mask) \
-    "b.n skip_args%=\n" \
-    ".word " UVISOR_TO_STRING(UVISOR_REGISTER_GATEWAY_MAGIC) "\n" \
-    ".word " UVISOR_TO_STRING(addr) "\n" \
-    ".word " UVISOR_TO_STRING(src_box) "_cfg_ptr\n" \
-    ".word " UVISOR_TO_STRING(UVISOR_OP_WRITE(op)) "\n" \
-    ".word " UVISOR_TO_STRING(mask) "\n" \
-    "skip_args%=:\n"
-
-#endif /* __CC_ARM or __GNUC__ */
+/** Both `uvisor_read` and `uvisor_write` can have 2 additional input arguments -
+ *  mask and operation. These macros select the right argument from
+ *  `__VA_ARGS__` and provide a sensible default if none is provided.
+ * @internal
+ */
+#define SELECT_MASK(...)       __UVISOR_MACRO_SELECT(_0, ##__VA_ARGS__, , , _MASK , , _DEFAULT_NULL)(__VA_ARGS__)
+#define SELECT_READ_OP(...)    __UVISOR_MACRO_SELECT(_0, ##__VA_ARGS__, , , _OP , , _DEFAULT_READ)(__VA_ARGS__)
```

Comment:
Note that the `@internal` only applies to `SELECT_MASK(...)`. You can group it (`@{`,`@}`).